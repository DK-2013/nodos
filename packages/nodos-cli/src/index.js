const { Command, flags } = require('@oclif/command');

class NodosCliCommand extends Command {
  async run() {
    const opt = this.parse(NodosCliCommand);
    const name = opt.flags.name || 'world';
    this.log(`hello ${name} from ./src/index.js`);
  }
}

NodosCliCommand.description = `Describe the command here
...
Extra documentation goes here
`;

NodosCliCommand.flags = {
  // add --version flag to show CLI version
  version: flags.version({ char: 'v' }),
  // add --help flag to show CLI version
  help: flags.help({ char: 'h' }),
  name: flags.string({ char: 'n', description: 'name to print' }),
};

module.exports = NodosCliCommand;
