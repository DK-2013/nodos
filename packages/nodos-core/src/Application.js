export default class Application {
  defaultOptions = { headers: {}, params: null };

  constructor({ fastify, config, router }) {
    this.fastify = fastify;
    this.config = config;
    this.router = router;
  }

  listen(...args) {
    return this.fastify.listen(...args);
  }

  get = url => this.request('GET', url)

  post = (url, options = {}) => this.request('POST', url, options)

  put = (url, options = {}) => this.request('PUT', url, options)

  patch = (url, options = {}) => this.request('PATCH', url, options)

  delete = url => this.request('DELETE', url)

  request(method, url, options) {
    const { params, headers } = { ...this.defaultOptions, ...options };

    return this.fastify.inject({
      method,
      url,
      headers,
      payload: params,
    });
  }
}
