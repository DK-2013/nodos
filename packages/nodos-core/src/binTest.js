import jest from 'jest';

export default (args = process.argv.slice(2)) => {
  jest.run(['--testPathPattern', '/__tests__/', ...args]);
};
