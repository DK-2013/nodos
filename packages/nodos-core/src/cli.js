import _ from 'lodash';
import yargs from 'yargs';
import { nodos } from '.';
import log from './logger';

export default async (args = process.argv.slice(2), options = {}) => {
  const {
    exitProcess = true,
    projectRoot = process.cwd(),
    container = {},
    done = _.noop, // FIXME: hack for testing purposes
  } = options;
  const parser = yargs(args);
  parser.exitProcess(exitProcess);
  // parser.fail(done.fail);
  parser.demandCommand();
  parser.recommendCommands();
  parser.strict();
  parser.showHelpOnFail(true);
  const nodosItem = _.get(container, 'nodos', nodos);
  const app = await nodosItem(projectRoot);
  // .epilog(help.trim())
  // .example(example.trim())
  // console.log(app);
  // console.log(app.commandBuilders);
  const commands = app.commandBuilders.map(build => build({ app, container, done }));
  console.log(commands);
  commands.forEach(c => parser.command(c));
  return parser.argv;
};
