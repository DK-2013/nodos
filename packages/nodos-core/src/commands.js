import _ from 'lodash';
import repl from 'repl';
import columnify from 'columnify';
import { nodos } from '.';
// import buildRoutes from './routes';

// export default (projectRoot, gulp) => {
//   gulp.task('default', (done) => {
//     console.log('hello from The Nodos!');
//     done();
//   });
//   gulp.task('server', async () => {
//     const app = await nodos(projectRoot);
//     app.listen(3000, () => {
//     });
//   });

//   gulp.task('routes', async () => {
//     const router = buildRoutes(projectRoot);
//     router.routes.forEach((route) => {
//       console.log(route.name, route.method);
//     });
//   });
// };

export default (container, done) => [
  {
    command: 'console',
    describe: 'run console',
    builder: {},
    handler: async (argv) => {
      const replItem = _.get(container, 'repl', repl);
      const nodosItem = _.get(container, 'nodos', nodos);
      const replServer = replItem.start({
        prompt: '> ',
      });
      const app = nodosItem(argv.projectRoot);
      replServer.context.app = app;
      done();
    },
  },
  {
    command: 'server',
    describe: 'run server',
    handler: async (argv) => {
      const nodosItem = _.get(container, 'nodos', nodos);
      const app = await nodosItem(argv.projectRoot);
      app.listen(3000, () => {
        done();
      });
    },
  },
  {
    command: 'routes',
    describe: 'display routes',
    handler: async (argv) => {
      const nodosItem = _.get(container, 'nodos', nodos);
      const print = _.get(container, 'print', console.log);
      const { router: { routes } } = await nodosItem(argv.projectRoot);
      if (_.isEmpty(routes)) {
        print(
          'You don\'t have any routes defined!\n\n'
        + 'Please add some routes in config/routes.yml.',
        );
      } else {
        const formattedRoutes = columnify(
          routes,
          {
            columns: ['method', 'url', 'middlewares'],
            config: {
              method: {
                headingTransform: () => 'Verb',
                dataTransform: _.toUpper,
              },
              url: {
                headingTransform: () => 'URI Pattern',
              },
              middlewares: {
                headingTransform: () => 'Middlewares',
                dataTransform: names => `[ ${names} ]`,
              },
            },
          },
        );
        print(formattedRoutes);
      }
      done();
    },
  },
];
