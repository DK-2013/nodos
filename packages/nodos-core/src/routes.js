import yml from 'js-yaml';
// import debug from 'debug';
import { promises as fs } from 'fs';
import Router from '@nodosjs/routing';

// const log = debug('nodos');

export default async (config) => {
  const rawData = await fs.readFile(config.paths.routes);
  const routesMap = yml.safeLoad(rawData);
  return new Router(routesMap);
};
