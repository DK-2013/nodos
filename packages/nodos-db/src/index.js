import commandBuilder from './commands';
import Db from './Db';

export default async (config) => {
  const db = new Db(config.db);
  await db.connect();
  return {
    commandBuilders: [commandBuilder],
    context: { db },
    hooks: {
      async close() {
        await db.close();
      },
    },
  };
};
