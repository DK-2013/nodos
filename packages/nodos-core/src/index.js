import path from 'path';
import fastify from 'fastify';
import fastifyErrorPage from 'fastify-error-page';
// import fastifySensible from 'fastify-sensible';
import pointOfView from 'point-of-view';
import pug from 'pug';
// import debug from 'debug';
import buildRouter from './routes';
import binNodos from './binNodos';
import binTest from './binTest';
import Application from './Application';
import Response from './http/Response';
import log from './logger';

const nodosEnv = process.env.NODOS_ENV || 'development';


const bin = { nodos: binNodos, test: binTest };
export { bin };

const fetchMiddleware = async (config, middlewareName) => {
  const middleware = await import(path.join(config.paths.middlewares, middlewareName))
    .catch(() => import(path.join(__dirname, 'middlewares', middlewareName)))
    .catch(() => import(middlewareName))
    .catch(() => {
      throw new Error(`Middleware '${middlewareName}' is absent.`);
    });
  return middleware.default;
};

const sendResponse = (response, reply) => {
  log(response);
  Object.keys(response.headers).forEach((name) => {
    reply.header(name, response.headers[name]);
  });

  switch (response.responseType) {
    case 'code':
      reply.code(response.code).send();
      break;
    case 'rendering':
      reply.view(response.template(), response.locals);
      break;
    case 'redirect':
      reply.code(response.code).redirect(response.redirectUrl);
      break;

    default:
      throw new Error(`Unknown responseType: ${response.responseType}`);
  }
};

const buildFastify = async (config, router) => {
  const app = fastify({
    logger: true,
  });
  // app.register(fastifySensible);
  if (nodosEnv === 'development') {
    app.register(fastifyErrorPage);
  }
  app.register(pointOfView, {
    engine: { pug },
    includeViewExtension: true,
    templates: config.paths.templates,
  });

  app.get('/', (request, reply) => {
    request.log.info('Some info about the current request');
    reply.send({ hello: 'world' });
  });

  // console.log(router);
  const promises = router.routes.map(async (route) => {
    const pathToHandler = path.join(config.paths.handlers, route.resourceName);
    const middlewaresPromises = route.middlewares.map(fetchMiddleware.bind(null, config));
    const middlewares = await Promise.all(middlewaresPromises);
    const opts = {
      beforeHandler: middlewares,
    };
    app[route.method](route.url, opts, async (request, reply) => {
      // FIXME: enable only for development environment
      // FIXME: enable only for handlers and they deps
      Object.keys(require.cache).forEach((item) => { delete require.cache[item]; });
      const handlers = await import(pathToHandler);
      const response = new Response({ templateDir: route.resourceName, templateName: route.name });
      await handlers[route.name](request, response);
      sendResponse(response, reply);
    });
  });
  await Promise.all(promises);

  return app;
};

const buildConfig = async (projectRootPath) => {
  const join = path.join.bind(null, projectRootPath);
  const config = {
    paths: {
      routes: join('config', 'routes.yml'),
      application: join('config', 'application'),
      config: join('config'),
      environments: join('config/environments'),
      templates: join('app', 'templates'),
      handlers: join('app', 'handlers'),
      middlewares: join('app', 'middlewares'),
    },
  };

  const configureForApp = await import(config.paths.application);
  const pathToConfigForCurrentStage = path.join(config.paths.environments, `${nodosEnv}.js`);
  const configureForStage = await import(pathToConfigForCurrentStage);
  // FIXME: remove default
  configureForApp.default(config);
  configureForStage.default(config);
  return config;
};

export const nodos = async (projectRootPath) => {
  const config = await buildConfig(projectRootPath);
  const router = await buildRouter(config);
  const fstf = await buildFastify(config, router);
  return new Application({ config, fastify: fstf, router });
};
