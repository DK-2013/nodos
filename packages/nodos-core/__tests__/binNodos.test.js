import path from 'path';
import { bin, nodos } from '../src';

test('nodos/binNodos/console', (done) => {
  const replServer = { context: {} };
  const container = {
    repl: {
      start: () => replServer,
    },
  };
  bin.nodos(
    ['--projectRoot', path.join(__dirname, '__fixtures__/app'), 'console'],
    { container, done, exitProcess: false },
  );

  expect(replServer.context).toHaveProperty('app');
});

test('nodos/binNodos/server', (done) => {
  const fillResult = (port, cb) => {
    cb();
  };
  const container = {
    nodos: () => Promise.resolve({ listen: fillResult }),
  };
  bin.nodos(
    ['--projectRoot', path.join(__dirname, '__fixtures__/app'), 'server'],
    { container, done, exitProcess: false },
  );
});

describe('nodos/binNodos/routes', () => {
  test('return valid routes presentation', (done) => {
    const container = {
      nodos: () => nodos(`${__dirname}/__fixtures__/app`),
      print: (output) => { expect(output).toMatchSnapshot(); },
    };
    bin.nodos(
      ['--projectRoot', path.join(__dirname, '__fixtures__/app'), 'routes'],
      { container, done },
    );
  });

  test('return valid presentation when no routes defined', (done) => {
    const container = {
      nodos: () => Promise.resolve({
        router: { routes: [] },
      }),
      print: (output) => { expect(output).toMatchSnapshot(); },
    };
    bin.nodos(
      ['--projectRoot', path.join(__dirname, '__fixtures__/app'), 'routes'],
      { container, done },
    );
  });
});
