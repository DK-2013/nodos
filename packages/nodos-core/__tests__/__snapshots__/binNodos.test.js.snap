// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nodos/binNodos/routes return valid presentatation when no routes defined 1`] = `
"You don't have any routes defined!

Please add some routes in config/routes.yml."
`;

exports[`nodos/binNodos/routes return valid routes presentation 1`] = `
"Verb   URI Pattern         Middlewares        
GET    /api/users          [ checkMigrations ]
GET    /api/users/new      [ checkMigrations ]
POST   /api/users          [ checkMigrations ]
GET    /api/users/:id      [ checkMigrations ]
GET    /api/users/:id/edit [ checkMigrations ]
PATCH  /api/users/:id      [ checkMigrations ]
PUT    /api/users/:id      [ checkMigrations ]
DELETE /api/users/:id      [ checkMigrations ]
GET    /users              [ checkMigrations ]
GET    /users/new          [ checkMigrations ]
POST   /users              [ checkMigrations ]
GET    /users/:id          [ checkMigrations ]
GET    /users/:id/edit     [ checkMigrations ]
PATCH  /users/:id          [ checkMigrations ]
PUT    /users/:id          [ checkMigrations ]
DELETE /users/:id          [ checkMigrations ]
GET    /session/new        [ checkMigrations ]
GET    /session            [ checkMigrations ]
GET    /session/edit       [ checkMigrations ]
PATCH  /session            [ checkMigrations ]
PUT    /session            [ checkMigrations ]
DELETE /session            [ checkMigrations ]
GET    /articles           [ checkMigrations ]
GET    /articles/new       [ checkMigrations ]
POST   /articles           [ checkMigrations ]
GET    /articles/:id       [ checkMigrations ]
GET    /articles/:id/edit  [ checkMigrations ]
PATCH  /articles/:id       [ checkMigrations ]
PUT    /articles/:id       [ checkMigrations ]
DELETE /articles/:id       [ checkMigrations ]"
`;
